docs/design/operator.md

```diff
@@ -0,0 +1,387 @@
+          hostPath:
+            path: "/run"
+            type: Directory
+        volumeMounts:
+          name: ''
+          readOnly: true
+          mountPath: "/run"
+          mountPropagation: Bidirectional
+      kubeletDirPath: "/var/lib/kubelet"
+      imagePullPolicy: IfNotPresent
+    provisioner:
+      priorityClassName: system-cluster-critical
+      labels:
+        app: provisioner
+      annotations:
+        k8s.v1.cni.cncf.io/networks: macvlan-conf-1
+      provisionerReplicas: 2
+      affinity:
+        nodeAffinity:
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+            - matchExpressions:
+              - key: storage
+                operator: In
+                values:
+                - node
+      tolerations:
+      - key: storage
+        operator: Exists
+      resources:
+        attacher:
+          limits:
+            cpu: '200'
+            memory: '500'
+          requests:
+            cpu: '100'
+            memory: '250'
+        snapshotter:
+          limits:
+            cpu: '200'
+            memory: '500'
+          requests:
+            cpu: '100'
+            memory: '250'
+        resizer:
+          limits:
+            cpu: '200'
+            memory: '500'
+          requests:
+            cpu: '100'
+            memory: '250'
+        provisioner:
+          limits:
+            cpu: '200'
+            memory: '500'
+          requests:
+            cpu: '100'
+            memory: '250'
+        omapGenerator:
+          limits:
+            cpu: '200'
+            memory: '500'
+          requests:
+            cpu: '100'
+            memory: '250'
+        liveness:
+          limits:
+            cpu: '200'
+            memory: '500'
+          requests:
+            cpu: '100'
+            memory: '250'
+        plugin:
+          limits:
+            cpu: '200'
+            memory: '500'
+          requests:
+            cpu: '100'
+            memory: '250'
+    liveness:
+      metricsPort: 8000
+    leaderElection:
+      leaseDuration: 100
+      renewDeadline: 100
+      retryPeriod: 10
+    deployCSIAddons: true
+    cephFsKernelClient: true
+    cephFsKernelMountOptions: ms_mode=secure
+status:
+  phase: Succeeded
+  reason: operator config successfully created
+```
+
+### CephCSIDriver CRD
+
+Manages the installation, lifecycle management, and configuration for CephFS,
+RBD, and NFS CSI drivers within namespaces.
+
+- Allows customization of driver settings on a per-driver basis.
+- Best practice dictates only a single instance of RBD/NFS/CephFS drivers
+  should be created in a single namespace. If there is a need to deploy a
+  second instance of a driver of a similar type, it should be deployed in a
+  different namespace.
+- Each CSI driver must have a unique name across all namespaces.
+
+```yaml
+---
+kind: CephCSIDriver
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: "<prefix>.<driver_type>.csi.ceph.com"
```

Comment:
I have some concerns about using the CR name field to encode config options like this.

On one hand, I understand that this avoids certain kinds of complexity.

On the other hand, this feels like it would be hard for users to understand.
Do I have to provide a prefix? Does it have to be unique? Can I **not** provide a prefix if I don't want to?

What are the available driver types I can use? This is to say that the only indication I had from this example that it was a cephfs-specific driver was from `labels.app=cephfs-plugin`.

Considering how users will view this and how this will be documented, I think this will be less clear and that this is the kind of thing I see users get wrong often. 

I would suggest something like this as an alternative:
1. metadata.name can be whatever the user wants, with a reasonable suggestion being something like, "platform-cephfs-driver"
2. `prefix` becomes a required spec field
3. `driverType` becomes a required spec field with an `enum` set of options -- this makes it more clear (without having to have inside-knowledge, what the driver type is)
4. the design doc should specify what validations the operator will take to ensure that the resultant driver name is unique in the k8s cluster
5. to support usability/user feedback, add `status.driverName` = `<prefix>.<driverType>.csi.ceph.com` so that users have feedback to understand what driver name to use in StorageClasses.

Reply:
>Do I have to provide a prefix? Does it have to be unique? Can I not provide a prefix if I don't want to?


This is not always required but that can be set but the format need to be followed.


> What are the available driver types I can use? This is to say that the only indication I had from this example that it was a cephfs-specific driver was from labels.app=cephfs-plugin.

This is just an example in the design doc and more specific CR will be provided when we have proper deployment guide.

>Considering how users will view this and how this will be documented, I think this will be less clear and that this is the kind of thing I see users get wrong often.

>// The name MUST follow domain name notation format
  // (https://tools.ietf.org/html/rfc1035#section-2.3.1). It SHOULD
  // include the plugin's host company name and the plugin name,
  // to minimize the possibility of collisions. It MUST be 63
  // characters or less, beginning and ending with an alphanumeric
  // character ([a-z0-9A-Z]) with dashes (-), dots (.), and
  // alphanumerics between. This field is REQUIRED.

we already through a couple of options before putting up the design doc with the same idea you have presented. we wanted to keep this CR the way how the csidriver looks today, the should be the unique name what the user wants the driver to the named. the CSI driver has some hard contstrain with the naming convention i mentioned above, Instead of introducing new field

with prefix and type in the spec it gives user an option to create duplicate csi driver easily with some which we are really trying to avoid to an extend by moving it to the name as the name of the CR is the name of the csi driver.

@nb-ohad anything to add?

Reply:
@BlaineEXE we do understand this is not a perfect solution but a decision based on "the best of two evils" approach. The requirements are as follows:

All driver names have to be unique! and conflict needs to be caught at creation time, not at reconciliation time. The name will be carried to the actual driver name as we want a 1:1 match between the client-exposed name at the operator layer and the name in the CSI implementation layer.

We dropped the type as a field in the spec for 2 reasons:
- I can create a CR that is called `my-rbd-driver` and have it of type `cephfs`
- Even if we agree that it is a user mistake, we cannot force the requirement to match the CSI driver implementation name to be `my-rbd-driver` because of CSI name restrictions 

On the other hand, if we encode the type in the name we can solve all the issues with some simple cel rules which gives the user a pretty good experience when it come to explanting the limitation when creation fails.